sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,148 @@
+import { PollOperation } from "./pollOperation";
+
+export type CancelOnProgress = () => void;
+
+export interface PollProgressSubscriber<TProperties, TResult> {
+  id: string;
+  conditional: (op?: PollOperation<TProperties, TResult>) => boolean;
+  callback: (op?: PollOperation<TProperties, TResult>) => void;
+}
+
+export abstract class Poller<TProperties, TResult> {
+  private stopped: boolean;
+  private resolve?: (value?: TResult) => void;
+  private reject?: (error: Error) => void;
+  private pollOncePromise?: Promise<void>;
+  private cancelPromise?: Promise<PollOperation<TProperties, TResult>>;
+  private promise: Promise<TResult>;
+  private pollProgressSubscribers: PollProgressSubscriber<TProperties, TResult>[] = [];
+  public operation: PollOperation<TProperties, TResult>;
+
+  constructor(operation: PollOperation<TProperties, TResult>, stopped: boolean = false) {
+    this.operation = operation;
+    this.stopped = stopped;
+    this.promise = new Promise(
+      (
+        resolve: (result?: TResult) => void,
+        reject: (error: Error) => void
+      ) => {
+        this.resolve = resolve;
+        this.reject = reject;
+      }
+    );
+    this.operation.state.started = true;
+    this.startPolling();
+  }
+
+  abstract async delay(): Promise<void>;
+  abstract async getResult(): Promise<TResult | undefined>;
+
+  private async startPolling(): Promise<void> {
+    while (!this.isStopped() && !this.isDone()) {
+      await this.poll();
+      await this.delay();
+    }
+  }
+
+  private async pollOnce(): Promise<void> {
+    try {
+      if (!this.isDone()) {
+        this.operation = await this.operation.update();
+        if (this.isDone() && this.resolve) {
+          const result = await this.getResult();
+          this.operation.state.result = result;
+          this.resolve(result);
+        }
+      }
+    } catch (e) {
+      this.operation.state.error = e;
+      if (this.reject) {
+        this.reject(e);
+      }
+    }
+  }
+
+  public poll(): Promise<void> {
+    if (!this.pollOncePromise) {
+      this.pollOncePromise = this.pollOnce();
+      this.pollOncePromise.finally(() => {
+        for (const subscriber of this.pollProgressSubscribers) {
+          if (subscriber.conditional(this.operation)) {
```

Comment:
Hmm this is an interesting design, since it puts the burden on the subscriber to decide if the state change is interesting or not. If that's the case, I'm not sure it matters to have two fields for this, since the subscriber could just ignore the state in the callback.

What I was thinking was something slightly different, such as passing a callback function to operation.update(), e.g. inside of pollOnce e.g.

```ts
pollOnce() {
   // ... other stuff
   this.operation = await this.operation.update(this.fireProgress);
   // ... other stuff
}

private fireProgress(properties: TProperties) {
   // fire to each subscriber function
}
```

This way the operation gets to decide when to notify the client, since it has the best idea of what changes are interesting.

Reply:
Gotcha! I'll do that then